src/sentry/middleware/ratelimit.py

```diff
@@ -19,42 +27,55 @@ class RatelimitMiddleware(MiddlewareMixin):
-                    status=429,
-                )
+        try:
+            # TODO: put these fields into their own object
+            request.will_be_rate_limited = False
+            request.rate_limit_category = None
+            request.rate_limit_uid = uuid.uuid4().hex
+            request.rate_limit_key = get_rate_limit_key(view_func, request)
+            if request.rate_limit_key is None:
+                return
+            category_str = request.rate_limit_key.split(":", 1)[0]
+            request.rate_limit_category = category_str
+
+            rate_limit = get_rate_limit_value(
+                http_method=request.method,
+                endpoint=view_func.view_class,
+                category=RateLimitCategory(category_str),
+            )
+            if rate_limit is None:
+                return
+
+            request.rate_limit_metadata = above_rate_limit_check(
+                request.rate_limit_key, rate_limit, request.rate_limit_uid
+            )
+
+            # TODO: also limit by concurrent window once we have the data
+            if request.rate_limit_metadata.rate_limit_type == RateLimitType.FIXED_WINDOW:
+                request.will_be_rate_limited = True
+                enforce_rate_limit = getattr(view_func.view_class, "enforce_rate_limit", False)
+                if enforce_rate_limit:
+                    return HttpResponse(
+                        {
+                            "detail": DEFAULT_ERROR_MESSAGE.format(
+                                limit=request.rate_limit_metadata.limit,
+                                window=request.rate_limit_metadata.window,
+                            )
+                        },
+                        status=429,
+                    )
+        except Exception:
+            logging.exception("Error during rate limiting, failing open. THIS SHOULD NOT HAPPEN")
 
     def process_response(self, request: Request, response: Response) -> Response:
-        if hasattr(request, "rate_limit_metadata"):
-            response["X-Sentry-Rate-Limit-Remaining"] = request.rate_limit_metadata.remaining
-            response["X-Sentry-Rate-Limit-Limit"] = request.rate_limit_metadata.limit
-            response["X-Sentry-Rate-Limit-Reset"] = request.rate_limit_metadata.reset_time
+        rate_limit_metadata: RateLimitMeta | None = getattr(request, "rate_limit_metadata", None)
+        if rate_limit_metadata:
+            response["X-Sentry-Rate-Limit-Remaining"] = rate_limit_metadata.remaining
+            response["X-Sentry-Rate-Limit-Limit"] = rate_limit_metadata.limit
+            response["X-Sentry-Rate-Limit-Reset"] = rate_limit_metadata.reset_time
+            response[
+                "X-Sentry-Rate-Limit-ConcurrentRemaining"
+            ] = rate_limit_metadata.concurrent_remaining
+            response["X-Sentry-Rate-Limit-ConcurrentLimit"] = rate_limit_metadata.concurrent_limit
+        finish_request(request.rate_limit_key, request.rate_limit_uid)
```

Comment:
If for someone reason, `finish_request` doesn't run (like from a crash), would the uid leak?

Reply:
yes it would, but it would be cleaned up within 30 seconds